server/impl/src/main/java/com/walmartlabs/concord/server/process/waits/WaitProcessFinishHandler.java

```diff
@@ -52,16 +75,127 @@ public WaitType getType() {
+
+        return Result.of(item.processKey(), item.waitConditionId(),
+                ProcessCompletionCondition.builder().from(item.waitCondition())
+                        .processes(newAwaitProcesses)
+                        .build());
+    }
+
+    private static WaitConditionItem<ProcessCompletionCondition> toModifiable(WaitConditionItem<ProcessCompletionCondition> item) {
+        return WaitConditionItem.of(item.processKey(), item.waitConditionId(), ModifiableProcessCompletionCondition.create().from(item.waitCondition()));
+    }
+
+    private static List<Set<UUID>> split(Set<UUID> processes, int maxProcessesItems) {
+        List<Set<UUID>> subsets = new ArrayList<>();
+        Set<UUID> currentSet = new HashSet<>();
+
+        for (UUID element : processes) {
+            currentSet.add(element);
+            if (currentSet.size() >= maxProcessesItems) {
+                subsets.add(new HashSet<>(currentSet));
+                currentSet.clear();
+            }
+        }
+
+        if (!currentSet.isEmpty()) {
+            subsets.add(currentSet);
+        }
+
+        return subsets;
+    }
+
+    private static void filterFinished(ProcessCompletionCondition condition, Map<UUID, ProcessStatus> statuses) {
+        if (condition.processes() == null || condition.processes().isEmpty()) {
+            return;
+        }
+
+        Iterator<UUID> it = condition.processes().iterator();
+        while (it.hasNext()) {
+            UUID proc = it.next();
+            if (statuses.containsKey(proc)) {
+                ProcessStatus status = statuses.get(proc);
+                if (status == null || condition.finalStatuses().contains(status)) {
+                    if (condition.completeCondition() == CompleteCondition.ALL) {
+                        it.remove();
+                    } else if (condition.completeCondition() == CompleteCondition.ONE_OF) {
+                        condition.processes().clear();
+                        return;
+                    } else {
+                        throw new RuntimeException("Unknown wait complete condition: " + condition.completeCondition());
+                    }
+                }
+            }
+        }
+    }
+
+    public static class Dao extends AbstractDao {
+
+        @Inject
+        private Dao(@MainDB Configuration cfg) {
+            super(cfg);
         }
 
-        if (wait.resumeEvent() != null) {
-            return Result.resume(wait.resumeEvent());
-        } else {
-            return Result.action(tx -> processQueueManager.updateExpectedStatus(tx, processKey, ProcessStatus.WAITING, ProcessStatus.ENQUEUED));
+        public Map<UUID, ProcessStatus> findStatuses(Set<UUID> processes) {
+            Map<UUID, ProcessStatus> result = txResult(tx -> {
+                ProcessQueue q = PROCESS_QUEUE.as("q");
+                return tx.select(q.INSTANCE_ID, q.CURRENT_STATUS)
+                        .from(q)
+                        .where(q.INSTANCE_ID.in(processes))
```

Comment:
Is JOOQ smart enough to turn this `INSTANCE_ID IN (...)` into an array operation? I'm worried that it's going to fail for large `processes` sets.

Reply:
that's why we split processes here:
https://github.com/walmartlabs/concord/pull/910/files/8a2282d40442a02df930f0192f4dd1f2b5e6e391#diff-33569c60af8758f4910c839fbc4b973bd155e2cc59016ca55f0d9cfe354948a3R90

Reply:
So the upper limit is `waitProcessLimitForStatusQuery`? Should be fine, I think, but for the query parser sake I'd replaced with an array operation if it is possible here.

Reply:
yes, `waitProcessLimitForStatusQuery` is a limit